### PR TITLE
Added Dex price volatility lesson to teach about oracle abuse, also added kovan information

### PR DESCRIPTION
**Pull request by PatrickAlphaC**
Hi all! Had a lot of fun learning the inner workings of Ethernaut to add a level here.

The idea is that flash loan attacks are rampant in the defi world, and having some experience with some of the price issues is going to be a must for up and coming solidity devs. 

basically, the Dex is a really crappy uniswap. It reports a price based on liquidity pools, and a user can use this to flip back and forth and eventually drain the dex of the liquidity pools. 

I was thinking a V2 might be using some sort of flash loan attack. It was a little tricky to involve the ERC20s due to the main interaction the player has with the world is through the `contract`, so I had to add a few functions to the main contract that allow to approve ERC20 interactions. 

If you can think of a better way to implement that, feel free! Also this gives the solidity devs more experience working with openzepplin packages :D 

Hope you like it! Let me know if I can do anything to make things nicer.

### CODE REVIEW

**Comment by abcoathup** [comment]
Hi @PatrickAlphaC,

Thanks for the PR.

Would you mind sharing in the community forum https://forum.openzeppelin.com to ask for community reviews?

**Comment by PatrickAlphaC** [comment]
100% :D

**Comment by xaler5** [comment]
Hey @PatrickAlphaC hi!

I've reviewed the PR and I would have a couple of suggestions:

- I've noticed you're using the `ERC20` contract instead of referring to the token contract with the `IERC20` interface. What do you think of using it wherever possible ?
- I think you forgot to run `npm install --save ethers` because if you don't install it manually, tests are failing. We need to have it added as dev dependency in the project since it is used exclusively on tests.
-  Users past progresses are in Rinkeby network. Although we would like to support more networks at some point, it would currently be difficult for us to maintain so we will remain on Rinkeby exclusively for now. Could you remove the Kovan deployment information?
- Regarding the `dex_completed` document. I see it interesting and I would like it to have it more general. There are different kind of oracles implementations, each one with its tradeoffs and strong points. It would be great to have a reference for a couple of them pointing to the fact that security is of outter importance when talking about oracles. That would be a great learning for the user playing this game. What do you think ?
- The variables names `tkn1` and `tkn2` are nice if the intent is to ocult to the player their intents and make his task harder, but I see that they are not aligned with the rest of the variables names used in the other levels of the game. We tend to be explicit in the names used and have an easier readability.

**Comment by PatrickAlphaC** [comment]
Hi @xaler5!

Thanks so much for the review!

I will make all the suggestions you have described here and update the PR. For the oracles part, I can a more general vibe for sure, and perhaps mention uniswap as another oracle choice?

**Comment by PatrickAlphaC** [comment]
Hi @xaler5,

I made the suggestions, and also cleaned up the testing sweet as well. 

One the completed section, happy to add more in there if you have some suggestions.

Looking forward to the feedback here!

**Comment by PatrickAlphaC** [comment]
Not sure why Travis is failing :(

File: gamedata/descriptions/levels/dex.md [hunk 1]
```diff
@@ -0,0 +1,15 @@
+The goal of this level is for you to hack the basic [DEX](https://en.wikipedia.org/wiki/Decentralized_exchange) contract below and steal the funds by price manipulation.
+
+You will start 10 tokens of `token1` and 10 of `token2`. The DEX contract starts with 100 of each token.
```
Comment (xaler5) [inline gamedata/descriptions/levels/dex.md:3]:
```suggestion
You will start with 10 tokens of `token1` and 10 of `token2`. The DEX contract starts with 100 of each token. 
```

File: gamedata/descriptions/levels/dex_completed.md [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+[Chainlink Data Feeds](https://docs.chain.link/docs/get-the-latest-price) are a secure, reliable, way to get decentralized data into your smart contracts. They have a vast library of many different sources, and also offer [secure randomness](https://docs.chain.link/docs/chainlink-vrf), ability to make [any API call](https://docs.chain.link/docs/make-a-http-get-request), [modular oracle network creation](https://docs.chain.link/docs/architecture-decentralized-model), [upkeep, actions, and maintainance](https://docs.chain.link/docs/kovan-keeper-network-beta), and unlimited customization. 
+
+[Uniswap TWAP Oracles](https://uniswap.org/docs/v2/core-concepts/oracles/) rely on [TWAP](https://en.wikipedia.org/wiki/Time-weighted_average_price#) and rely solely on the liquidity of the DEX protocol.
```
Comment (xaler5) [inline gamedata/descriptions/levels/dex_completed.md:9]:
```suggestion
[Uniswap TWAP Oracles](https://uniswap.org/docs/v2/core-concepts/oracles/) relies on a time weighted price model called [TWAP](https://en.wikipedia.org/wiki/Time-weighted_average_price#). While the design can be attractive, this protocol heavily depends on the liquidity of the DEX protocol, and if this is too low, prices can be easily manipulated. 
```

**Comment by xaler5** [comment]
Hey @PatrickAlphaC thanks for the changes, I just added to little changes in phrasing and wording.

@frangio I think this is good to go

**Comment by PatrickAlphaC** [comment]
Much cleaner wording, great call. Is there anything else we think makes sense?

**Comment by frangio** [comment]
@PatrickAlphaC @xaler5 I removed the dependency on Ethers (this project has too many dependencies already :zipper_mouth_face:) which I think will help with the Travis failure.

I've also moved the level to the end of the list, mostly because we need the `deployId` of the other levels to remain the same (they had been incremented) and I'm honestly not so familiar with these deployment scripts and I'm not sure if anything else would work.

Lastly I removed what I believe were some unnecessary changes to the JSON files.

Thank you @PatrickAlphaC!

**Comment by PatrickAlphaC** [comment]
Awesome! Thanks @frangio and @xaler5!

I guess this means this will go at the end of the course yes? And whoops, forgot to remove the rest of the kovan pieces!

In any case, this is awesome!

**Review by xaler5 — APPROVED** [review]

**Comment by frangio** [comment]
Yes, it will go at the end of the course for now. We can review the order but we need to see if that affects anything else.

**Comment by xaler5** [comment]
I'll merge and deploy on rinkeby the level then. Thanks everyone

**Comment by MikeBailleul** [comment]
The approve() function on the contract is actually a security issue.
You can hack the DEX and drain the pool by giving allowance from ANY address without doing any price manipulation.

Like so:
`contract Attacker {
    
    SwappableToken public token1;

    constructor(address _token1Addr) public {
        token1 = SwappableToken(_token1Addr);
    }

    function attack(address dex) public {
        token1.approve(dex, address(this), 100);
        token1.transferFrom(dex, address(this), 100);
    }

}`

Suggested fix:
`function approve(address spender, uint amount) public returns(bool){
        super._approve(msg.sender, spender, amount);
  }`

**Comment by PatrickAlphaC** [comment]
Showing how easy it is to make a bad contract :)

I tried to be clever to make the game a little more easier for people. We had to take some liberties to "gamify" the issue. But you're right, we shouldn't have the contract setup in a way where there is a way people _should_ solve it.

**Comment by PatrickAlphaC** [comment]
https://github.com/OpenZeppelin/ethernaut/pull/225 Updated

**Pull request merged**